src/dbnode/storage/series/buffer.go

```diff
@@ -500,18 +475,205 @@ func (b *dbBuffer) FetchBlocksMetadata(
 			LastRead: resultLastRead,
 		})
-	})
+	}
 
 	return res
 }
 
-type dbBufferBucket struct {
-	opts              Options
+func (b *dbBuffer) newBucketsAt(
+	t time.Time,
+) *dbBufferBuckets {
+	buckets := b.bucketsPool.Get()
+	buckets.resetTo(t, b.opts, b.bucketPool)
+	b.bucketsMap[xtime.ToUnixNano(t)] = buckets
+	return buckets
+}
+
+func (b *dbBuffer) bucketsAt(
+	t time.Time,
+) (*dbBufferBuckets, bool) {
+	// First check LRU cache
+	for _, buckets := range b.bucketsCache {
+		if buckets == nil {
+			continue
+		}
+		if buckets.start.Equal(t) {
+			return buckets, true
+		}
+	}
+
+	// Then check the map
+	if buckets, exists := b.bucketsMap[xtime.ToUnixNano(t)]; exists {
+		return buckets, true
+	}
+
+	return nil, false
+}
+
+func (b *dbBuffer) bucketsAtCreate(
+	t time.Time,
+) *dbBufferBuckets {
+	if buckets, exists := b.bucketsAt(t); exists {
+		return buckets
+	}
+
+	return b.newBucketsAt(t)
+}
+
+func (b *dbBuffer) putBucketInCache(newBuckets *dbBufferBuckets) {
+	replaceIdx := bucketsCacheSize - 1
+	for i, buckets := range b.bucketsCache {
+		// Check if we have the same pointer in cache
+		if buckets == newBuckets {
+			replaceIdx = i
+		}
+	}
+
+	for i := replaceIdx; i > 0; i-- {
+		b.bucketsCache[i] = b.bucketsCache[i-1]
+	}
+
+	b.bucketsCache[0] = newBuckets
+}
+
+func (b *dbBuffer) removeBucketsAt(blockStart time.Time) {
+	buckets, exists := b.bucketsAt(blockStart)
+	if !exists {
+		return
+	}
+
+	b.bucketsPool.Put(buckets)
+	delete(b.bucketsMap, xtime.ToUnixNano(blockStart))
+}
+
+type dbBufferBuckets struct {
+	buckets           []*dbBufferBucket
 	start             time.Time
-	encoders          []inOrderEncoder
-	bootstrapped      []block.DatabaseBlock
+	opts              Options
 	lastReadUnixNanos int64
-	drained           bool
+	bucketPool        *dbBufferBucketPool
+}
+
+func (b *dbBufferBuckets) resetTo(
+	start time.Time,
+	opts Options,
+	bucketPool *dbBufferBucketPool,
+) {
+	// nil all elements so that they get GC'd
+	for i := range b.buckets {
+		b.buckets[i] = nil
+	}
+	b.buckets = b.buckets[:0]
+	b.start = start
+	b.opts = opts
+	atomic.StoreInt64(&b.lastReadUnixNanos, 0)
+	b.bucketPool = bucketPool
+}
+
+func (b *dbBufferBuckets) streams(ctx context.Context, wType WriteType) []xio.BlockReader {
+	var res []xio.BlockReader
+	for _, bucket := range b.buckets {
+		if bucket.wType == wType {
+			res = append(res, bucket.streams(ctx)...)
+		}
+	}
+
+	return res
+}
+
+func (b *dbBufferBuckets) streamsLen() int {
+	res := 0
+	for _, bucket := range b.buckets {
+		res += bucket.streamsLen()
+	}
+	return res
+}
+
+func (b *dbBufferBuckets) write(
+	timestamp time.Time,
+	value float64,
+	unit xtime.Unit,
+	annotation []byte,
+	wType WriteType,
+) error {
+	return b.writableBucketCreate(wType).write(timestamp, value, unit, annotation)
+}
+
+func (b *dbBufferBuckets) merge(wType WriteType) (mergeResult, error) {
+	var res mergeResult
+	for _, bucket := range b.buckets {
+		// Only makes sense to merge buckets that are writable
```

Comment:
why? Because the other ones have been flushed and will be evicted soon?

Reply:
Non-writable buckets are guaranteed to be merged and not written to after that.